src/hotspot/share/gc/g1/heapRegionSet.cpp

```diff
@@ -236,31 +236,33 @@ void FreeRegionList::remove_starting_at(HeapRegion* first, uint num_regions) {
   DEBUG_ONLY(uint old_length = length();)
 
   HeapRegion* curr = first;
+  HeapRegion* prev = curr->prev();
+  if (prev == NULL) {
+    assert_free_region_list(_head == curr, "invariant");
+  } else {
+    assert_free_region_list(_head != curr, "invariant");
+  }
+  HeapRegion* next = curr->next();
+  if (next == NULL) {
+    assert_free_region_list(_tail == curr, "invariant");
+  } else {
+    assert_free_region_list(_tail != curr, "invariant");
+  }
   uint count = 0;
   while (count < num_regions) {
     verify_region(curr);
-    HeapRegion* next = curr->next();
-    HeapRegion* prev = curr->prev();
+    next = curr->next();
 
+    assert_free_region_list(_head != next, "invariant");
+    if (next != NULL) {
+      assert_free_region_list(next->prev() != NULL, "invariant");
+      assert_free_region_list(_tail != curr, "invariant");
+    }
     assert(count < num_regions,
```

Comment:
Since you are touching this area, `assert(count < num_regions)` can be dropped. (We are inside the while-loop; this condition must hold.) A more useful assert is sth like `length() >= num_regions`, if I understand the original code correctly. A bit surprised to not this mentioned in the doc.